src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/speaker/SpeakerRouter.java

```diff
@@ -157,7 +163,12 @@ private void proxySpeaker(Tuple input, InfoData payload) throws PipelineExceptio
             emit(STREAM_PORT_ID, input, makePortTuple(
                     new UpdatePortPropertiesCommand((UpdatePortPropertiesRequest) payload)));
         } else if (payload instanceof DeactivateIslInfoData) {
-            emit(STREAM_ISL_ID, input, makeIslTuple(input, new IslDeleteCommand((DeactivateIslInfoData) payload)));
+            DeactivateIslInfoData data = (DeactivateIslInfoData) payload;
+            emit(STREAM_ISL_ID, input, makeIslTuple(input, new IslDeleteCommand(data)));
+            emit(STREAM_UNI_ISL_ID, input,
+                    makeUniIslTuple(input, new UniIslSetupCommand(new Endpoint(data.getSource()))));
```

Comment:
Em... you reuse setup command to remove data. As for me such code reuse... makes more harm than good. At least you will have an extreme mess in logs - there is no person who will treat the message `"Uni-ISL service receive SETUP request for {}"` as "remove" of uni-isl data.

One more even bigger issue here - you "clean" uniIsl without any checks... IslFsm can reject or postpone removal, but you remove any way to communicate to it in advance, so all ISL clean logic becomes corrupted. You must clean data in cascade and only when the current step is done you can go to the next step.

So - only IslFsm/service can call further cleanup when it has done with its own cleanup. The easiest way here is to make the switch to the exhausted discovery mode from the Isl side as part of the removal process.

Reply:
You're right. I've added the switch to exhausted mode to ISL FSM.